### PR TITLE
cs-4687 update to get list of tasks from parameters

### PR DESCRIPTION
**Pull request by tintinthong**
- **p1**: Updating call to get rules from within the "tasks" attribute
- **p2**: Flattening "explanation" when scheduler writes the parameters to the s3 file to take on the new rule structure
- **p3**: Adding metadata and explanation into rule abstraction so it can be easily accessed
- **p4**: Passing row-by-row explanation data as a map column in parquet format. Reason why this is not done at the payment list level is to avoid roots changing due to explanation
- **p5**: Added a function to the rule abstraction that will basically get row-by-row explanation data given payment list. For other rules I put a placeholder first. 
- **p6**: If you want to see the parameters file in s3 before the task is run. I have updated the parameters file. It is flattening the rule structure but for each rule.

Ps: Im doing a separate pr for rollover. also implementation of staking in the rollover pr  

**Example of new rule structure**

```
{
  "explanation": {
    "en": {
      "program": "Reward Program managed by Cardstack Foundation",
      "claim": {
        "flat_payment": "Earned {amount} {token} for a one-time aidrop",
        "retro_airdrop": "Earned {amount} {token} averaged over time period for payments within {start_snapshot_block} and {end_snapshot_block}",
        "safe_ownership": "Earned {amount} {token} for owning a {safe_type}",
        "staking": "Earned {amount} {token} for staking CARD monthly in your L2 depot between blocks {from_block} and {to_block}, {rollover_amount} tokens rolled over from unclaimed previous cycles",
        "profile_creation": "Earned {amount} {token} for creating first profile by block {from_block}"
      }
    }
  },
  "tasks": [
    {
      "core": {
        "payment_cycle_length": 524000,
        "start_block": 17265697,
        "end_block": 23693329,
        "subgraph_config_locations": {
          "token_holder": "s3://cardpay-production-partitioned-graph-data/data/token_holder/1.0.0",
          "safe_owner": "s3://cardpay-production-partitioned-graph-data/data/safe_owner/0.0.1"
        },
        "duration": 43200
      },
      "user_defined": {
        "token": "0x52031d287Bb58E26A379A7Fec2c84acB54f54fe3",
        "duration": 43200,
        "interest_rate_monthly": 0.0048
      }
    }
  ]
}

```

### CODE REVIEW

File: packages/cardpay-reward-scheduler/reward_scheduler/reward_program.py [hunk 1]
```diff
@@ -131,7 +131,12 @@ def get_rules(self):
         if blob and blob != b"":
             try:
                 did = blob.decode("utf-8")  # new blob format: hex encodes a did string
-                rules = resolve_rule(did)
+                doc = resolve_did(did)
+                rules = doc.get("tasks", None)
+                if rules is None:
+                    raise Exception("No rules specified")
+                else:
+                    return rules
```
Comment (tintinthong) [inline packages/cardpay-reward-scheduler/reward_scheduler/reward_program.py:139]: p1

File: packages/cardpay-reward-scheduler/reward_scheduler/reward_program.py [hunk 2]
```diff
@@ -245,14 +260,15 @@ def run_all_rules(self) -> None:
 
         self.update_processed()
         rules = self.get_rules()
+        explanation_block = rules.get("explanation", {})
         logging.info(f"Reward program {self.reward_program_id} has {len(rules)} rules")
         self.raise_on_payment_cycle_overlap(rules)
         for rule in rules:
-            self.run_rule(rule)
+            self.run_rule(rule, explanation_block)
```
Comment (tintinthong) [inline packages/cardpay-reward-scheduler/reward_scheduler/reward_program.py:267]: p2

File: packages/cardpay-reward-programs/cardpay_reward_programs/main.py [hunk 3]
```diff
@@ -53,12 +53,19 @@ def run_reward_program(
...
+                parameters["explanation"],
+                parameters["metadata"],
+            )
     payment_list = rule.get_payments(**parameters["run"]).to_dict("records")
-    tree = PaymentTree(payment_list, run_parameters=parameters)
-    table = tree.as_arrow()
+    tree = PaymentTree(payment_list, parameters)
+    explanation_data_arr = rule.get_explanation_data_arr(payment_list)
+    explanation_id = rule.explanation_id
+    table = tree.as_arrow(explanation_id, explanation_data_arr)
```
Comment (tintinthong) [inline packages/cardpay-reward-programs/cardpay_reward_programs/main.py:66]: p3

File: packages/cardpay-reward-programs/cardpay_reward_programs/rule.py [hunk 4]
```diff
@@ -10,9 +10,17 @@ class Rule(ABC):
...
+        self,
+        core_parameters,
+        user_defined_parameters,
+        explanation_block={},
+        metadata={},
+    ):
         self.connection = duckdb.connect(":memory:")
         self.set_core_parameters(**core_parameters)
+        self.set_metadata_parameters(**metadata)
+        self.set_explanation_parameters(explanation_block)
```
Comment (tintinthong) [inline packages/cardpay-reward-programs/cardpay_reward_programs/rule.py:23]: p3

File: packages/cardpay-reward-programs/cardpay_reward_programs/rules/flat_payment.py [hunk 5]
```diff
@@ -34,3 +42,11 @@ def run(self, payment_cycle: int, reward_program_id: str):
...
             for account in self.accounts
         )
+
+    def get_explanation_data_arr(self, payment_list):
+        explanation_data_arr = []
+        for payment in payment_list:
+            explanation_data_arr.append(
+                {"amount": payment["amount"], "token": payment["token"]}
+            )
+        return explanation_data_arr
```
Comment (tintinthong) [inline packages/cardpay-reward-programs/cardpay_reward_programs/rules/flat_payment.py:52]: p5

File: packages/cardpay-reward-programs/input/flat_payment/parameters.json [hunk 6]
```diff
@@ -1,25 +1,42 @@
...
+  "explanation": {
+    "en": {
+      "program": "Reward Program managed by Cardstack Foundation",
+      "claim": {
+        "flat_payment": "Earned {amount} {token} for a one-time aidrop",
+        "retro_airdrop": "Earned {amount} {token} averaged over time period for payments within {start_snapshot_block} and {end_snapshot_block}",
+        "safe_ownership": "Earned {amount} {token} for owning a {safe_type}",
+        "staking": "Earned {amount} {token} for staking CARD monthly in your L2 depot between blocks {from_block} and {to_block}, {rollover_amount} tokens rolled over from unclaimed previous cycles",
+        "profile_creation": "Earned {amount} {token} for creating first profile by block {from_block}"
+      }
```
Comment (tintinthong) [inline packages/cardpay-reward-programs/input/flat_payment/parameters.json:11]: p6

**Review by IanCal — CHANGES_REQUESTED** [review]

**Review by IanCal — APPROVED** [review]

**Pull request merged**